### ISSUE TITLE
False positive for BlockArguments and do blocks

### ISSUE THREAD

**Issue by amesgen** [ndmitchell/hlint#837]
```haskell
{-# LANGUAGE BlockArguments #-}

main = print do 17 + 25
```
results in
```
test.hs:3:14: Warning: Redundant do
Found:
  do 17 + 25
Perhaps:
  17 + 25

1 hint
```
which is invalid.

Side note: It's very nice that hlint now uses a parser which supports "modern" extensions like BlockArguments, great work!

**Comment by ndmitchell** [comment]
Thanks for the report. I've fixed this so the above example works. I don't quite have the exact heuristics where BlockArguments makes a singleton `do` sensible, but I've guessed if the parent is an App and the child isn't atomic then it's redundant.

**Comment by ocharles** [comment]
Could we also have this for infix applications? The following still (incorrectly) gets a hint:

```
          rows
            <&> do \( Rel8.Col bomId, ( Rel8.Col subId, Rel8.Col partId_ ) ) ->
                     Map.singleton bomId [ ( subId, partId_ ) ]
             &  Map.unionsWith (<>)
```

I can open a new issue if you'd prefer.

**Comment by ndmitchell** [comment]
OK, I think I've come to the realisation that redundant do isn't a monad hint (as I'd thought), but secretly a form of bracketing! Using the `needBracket` stuff HLint has for bracketing all the above examples work, so I've switched to that and hopefully got the general principle, and thus all the relevant cases.
